File: src/verification.ts
import { InvalidArgument } from '@tanker/errors';
import { assertNotEmptyString } from './types';

export type EmailVerificationMethod = { type: 'email'; email: string };
export type PassphraseVerificationMethod = { type: 'passphrase' };
export type KeyVerificationMethod = { type: 'verificationKey' };
export type OIDCVerificationMethod = {
  type: 'oidcIdToken';
  providerId: string;
  providerDisplayName: string;
};
export type PhoneNumberVerificationMethod = {
  type: 'phoneNumber';
  phoneNumber: string;
};
export type PreverifiedEmailVerificationMethod = {
  type: 'preverifiedEmail';
  preverifiedEmail: string;
};
export type PreverifiedPhoneNumberVerificationMethod = {
  type: 'preverifiedPhoneNumber';
  preverifiedPhoneNumber: string;
};
export type E2ePassphraseVerificationMethod = { type: 'e2ePassphrase' };

export type VerificationMethod =
  | EmailVerificationMethod
  | PassphraseVerificationMethod
  | KeyVerificationMethod
  | OIDCVerificationMethod
  | PhoneNumberVerificationMethod
  | PreverifiedEmailVerificationMethod
  | PreverifiedPhoneNumberVerificationMethod
  | E2ePassphraseVerificationMethod;

export type EmailVerification = { email: string; verificationCode: string };
export type PassphraseVerification = { passphrase: string };
export type KeyVerification = { verificationKey: string };
export type OIDCAuthorizationCodeVerification = {
  oidcProviderId: string;
  oidcAuthorizationCode: string;
  oidcState: string;
};
export type OIDCVerification = { oidcIdToken: string };
export type PhoneNumberVerification = {
  phoneNumber: string;
  verificationCode: string;
};
export type PreverifiedEmailVerification = { preverifiedEmail: string };
export type PreverifiedPhoneNumberVerification = {
  preverifiedPhoneNumber: string;
};
export type PreverifiedOidcVerification = {
  preverifiedOidcSubject: string;
  oidcProviderId: string;
};
export type E2ePassphraseVerification = { e2ePassphrase: string };

export type Verification =
  | EmailVerification
  | PassphraseVerification
  | KeyVerification
  | OIDCAuthorizationCodeVerification
  | OIDCVerification
  | PhoneNumberVerification
  | PreverifiedEmailVerification
  | PreverifiedPhoneNumberVerification
  | PreverifiedOidcVerification
  | E2ePassphraseVerification;

export type VerificationOptions = {
  withSessionToken?: boolean;
  allowE2eMethodSwitch?: boolean;
};

const validMethods = [
  'email',
  'passphrase',
  'verificationKey',
  'oidcAuthorizationCode',
  'oidcIdToken',
  'phoneNumber',
  'preverifiedEmail',
  'preverifiedPhoneNumber',
  'preverifiedOidcSubject',
  'e2ePassphrase',
];
const validKeys = [
  ...validMethods,
  'verificationCode',
  'oidcProviderId',
  'oidcState',
];
const validVerifOptionsKeys = ['withSessionToken', 'allowE2eMethodSwitch'];

export const assertVerification = (verification: Verification) => {
  if (!verification || typeof verification !== 'object')
    throw new InvalidArgument('verification', 'object', verification);

  if (Object.keys(verification).some((k) => !validKeys.includes(k)))
    throw new InvalidArgument(
      'verification',
      `should only contain keys in ${JSON.stringify(validKeys)}`,
      verification
    );

  const methodCount = validMethods.reduce(
    (count, key) => count + (key in verification ? 1 : 0),
    0
  );

  if (methodCount !== 1)
    throw new InvalidArgument(
      'verification',
      `should contain a single verification method in ${JSON.stringify(
        validMethods
      )}`,
      verification
    );

  if ('email' in verification) {
    assertNotEmptyString(verification.email, 'verification.email');
    if (!('verificationCode' in verification)) {
      throw new InvalidArgument(
        'verification',
        'email verification should also have a verificationCode',
        verification
      );
    }
    assertNotEmptyString(
      verification.verificationCode,
      'verification.verificationCode'
    );
  } else if ('passphrase' in verification) {
    assertNotEmptyString(verification.passphrase, 'verification.passphrase');
  } else if ('verificationKey' in verification) {
    assertNotEmptyString(
      verification.verificationKey,
      'verification.verificationKey'
    );
  } else if ('oidcAuthorizationCode' in verification) {
    assertNotEmptyString(
      verification.oidcProviderId,
      'verification.oidcProviderId'
    );
    assertNotEmptyString(
      verification.oidcAuthorizationCode,
      'verification.oidcAuthorizationCode'
    );
    assertNotEmptyString(verification.oidcState, 'verification.oidcState');
  } else if ('oidcIdToken' in verification) {
    assertNotEmptyString(verification.oidcIdToken, 'verification.oidcIdToken');
  } else if ('phoneNumber' in verification) {
    assertNotEmptyString(verification.phoneNumber, 'verification.phoneNumber');
    if (!('verificationCode' in verification)) {
      throw new InvalidArgument(
        'verification',
        'phoneNumber verification should also have a verificationCode',
        verification
      );
    }
    assertNotEmptyString(
      verification.verificationCode,
      'verification.verificationCode'
    );
  } else if ('preverifiedEmail' in verification) {
    assertNotEmptyString(
      verification.preverifiedEmail,
      'verification.preverifiedEmail'
    );
  } else if ('preverifiedPhoneNumber' in verification) {
    assertNotEmptyString(
      verification.preverifiedPhoneNumber,
      'verification.preverifiedPhoneNumber'
    );
  } else if ('preverifiedOidcSubject' in verification) {
    assertNotEmptyString(
      verification.preverifiedOidcSubject,
      'verification.preverifiedOidcSubject'
    );
    if (!('oidcProviderId' in verification)) {
      throw new InvalidArgument(
        'verification',
        'oidc pre-verification should also have a oidcProviderId',
        verification
      );
    }
    assertNotEmptyString(
      verification.oidcProviderId,
      'verification.oidcProviderId'
    );
  } else if ('e2ePassphrase' in verification) {
    assertNotEmptyString(
      verification.e2ePassphrase,
      'verification.e2ePassphrase'
    );
  }
};

export const assertVerificationOptions = (options?: VerificationOptions) => {
  if (!options) return;

  // noinspection SuspiciousTypeOfGuard
  if (typeof options !== 'object' || options instanceof Array) {
    throw new InvalidArgument('options', 'object', options);
  }

  if (Object.keys(options).some((k) => !validVerifOptionsKeys.includes(k)))
    throw new InvalidArgument(
      'options',
      `should only contain keys in ${JSON.stringify(validVerifOptionsKeys)}`,
      options
    );

  if (
    'withSessionToken' in options &&
    typeof options.withSessionToken !== 'boolean'
  )
    throw new InvalidArgument(
      'options',
      'withSessionToken must be a boolean',
      options
    );

  if (
    'allowE2eMethodSwitch' in options &&
    typeof options.allowE2eMethodSwitch !== 'boolean'
  )
    throw new InvalidArgument(
      'options',
      'allowE2eMethodSwitch must be a boolean',
      options
    );
};

File: android/src/main/java/com/tanker/clientreactnative/Verification.kt
package com.tanker.clientreactnative

import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeMap
import io.tanker.api.*
import io.tanker.bindings.TankerVerification

// This is really a package-method used as an "extension constructor" ersatz
fun Verification(json: ReadableMap): Verification {
    json.getString("email")?.let {
        return EmailVerification(it, json.getString("verificationCode")!!)
    }
    json.getString("passphrase")?.let {
        return PassphraseVerification(it)
    }
    json.getString("verificationKey")?.let {
        return VerificationKeyVerification(it)
    }
    json.getString("oidcAuthorizationCode")?.let { oidcAuthorizationCode ->
        return OIDCAuthorizationCodeVerification(
            json.getString("oidcProviderId")!!,
            oidcAuthorizationCode,
            json.getString("oidcState")!!
        )
    }
    json.getString("oidcIdToken")?.let {
        return OIDCIDTokenVerification(it)
    }
    json.getString("phoneNumber")?.let {
        return PhoneNumberVerification(it, json.getString("verificationCode")!!)
    }
    json.getString("preverifiedEmail")?.let {
        return PreverifiedEmailVerification(it)
    }
    json.getString("preverifiedPhoneNumber")?.let {
        return PreverifiedPhoneNumberVerification(it)
    }
    json.getString("e2ePassphrase")?.let {
        return E2ePassphraseVerification(it)
    }
    throw AssertionError("Invalid verification JS object received, check Typescript definitions match")
}

fun VerificationOptions(json: ReadableMap?): VerificationOptions {
    val options = VerificationOptions()
    if (json == null)
        return options
    if (json.hasKey("withSessionToken"))
        options.withSessionToken(json.getBoolean("withSessionToken"))
    if (json.hasKey("allowE2eMethodSwitch"))
        options.allowE2eMethodSwitch(json.getBoolean("allowE2eMethodSwitch"))
    return options
}

fun VerificationMethod.toWritableMap(): WritableMap {
    val json = WritableNativeMap()
    when (this) {
        is EmailVerificationMethod -> {
            json.putString("type", "email")
            json.putString("email", this.email)
        }
        is VerificationKeyVerificationMethod -> json.putString("type", "verificationKey")
        is PassphraseVerificationMethod -> json.putString("type", "passphrase")
        is OIDCIDTokenVerificationMethod -> json.putString("type", "oidcIdToken")
        is PhoneNumberVerificationMethod -> {
            json.putString("type", "phoneNumber")
            json.putString("phoneNumber", this.phoneNumber)
        }
        is PreverifiedEmailVerificationMethod -> {
            json.putString("type", "preverifiedEmail")
            json.putString("preverifiedEmail", this.preverifiedEmail)
        }
        is PreverifiedPhoneNumberVerificationMethod -> {
            json.putString("type", "preverifiedPhoneNumber")
            json.putString("preverifiedPhoneNumber", this.preverifiedPhoneNumber)
        }
        is E2ePassphraseVerificationMethod -> json.putString("type", "e2ePassphrase")
    }
    return json
}
